src/com/facebook/buck/jvm/groovy/GroovycStep.java

```diff
@@ -111,11 +118,32 @@ public String getDescription(ExecutionContext context) {
         .add(outputDirectory.toString());
     addCrossCompilationOptions(command);
 
-    command.addAll(extraArguments.or(ImmutableList.<String>of()))
-           .addAll(transform(sourceFilePaths, toStringFunction()));
+    command.addAll(extraArguments.or(ImmutableList.<String>of()));
+
+    command.add("@" + pathToSrcsList);
+
     return command.build();
   }
 
+  private void writePathToSourcesList(ImmutableList<Path> expandedSources) throws IOException {
+    filesystem.writeLinesToPath(
+        FluentIterable.from(expandedSources)
+            .transform(toStringFunction())
+            .transform(Javac.ARGFILES_ESCAPER),
+        pathToSrcsList);
+  }
+
+  private ImmutableList<Path> getExpandedSourcePaths(ImmutableSet<Path> javaSourceFilePaths) {
+    ImmutableList.Builder<Path> sources = ImmutableList.builder();
+    for (Path path : javaSourceFilePaths) {
+      String pathString = path.toString();
+      if (pathString.endsWith(".java") || pathString.endsWith(".groovy")) {
```

Comment:
We are the sources filtered at this stage? They should be sorted much earlier so we don't end up rebuilding stuff when sources that don't contribute change.

Why would it even be possible to specify sources that don't get compiled?